add blog for develop guide on how multiple data source to integrate w…

zhyuanqi opened the pull request:
### Description
add blog for develop guide on how multiple data source to integrate with dashboards plugin
 
### Issues Resolved
* https://github.com/opensearch-project/project-website/issues/2815
* https://app.asana.com/0/0/1207209331528869/f
* 
### Check List
- [ ] Commits are signed per the DCO using --signoff


By submitting this pull request, I confirm that my contribution is made under the terms of the BSD-3-Clause License.

seraphjiang commented:
Thanks @zhyuanqi 

@kavilla  @BionIT @zengyan-amazon would you help to review

kavilla commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,258 @@
+
+Before getting started, ensure you have set up a local development environment for OpenSearch Dashboards following our developer guide.  Once your Dashboards instance is up and running, you can generate your own Dashboards plugin using the OpenSearch Dashboards Plugin Generator by running the CLI command below:
+```
+node scripts/generate_plugin --name my_plugin_name
````
Unrelated to this pr. But perhaps we can consider updating the cli to include a question to ask if the developers wants to setup their plugin for osd?

zhyuanqi replied:
I think it includes multiple questions includes what kind of internal plugin and what the part will be. 
```
? Will this plugin be part of the OpenSearch Dashboards repository? Yes
? What type of internal plugin would you like to create (Use arrow keys)
❯ OpenSearch Dashboards Example 
  OpenSearch Dashboards OSS 
  OpenSearch Dashboards OSS Functional Testing 
```

kavilla commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,258 @@
+## 4 FAQ 
+
+Is any UI component provided by dataSourceManagemet used in core?
+Data source selector is now used in devtools, tutorial, saved object management page. You can see the component after enabling data_source.enabed in opensearch_dashbards.yml.
```
data_source.enabled*

zhyuanqi replied:
Got it.

kavilla reviewed (approved):
Looks great ! Couple comments but really just nitpicks. But I don't have the best grammar skills.

General nitpicks:
* I noticed it switches between "multiple data source" and "multiple data sources" it think the latter sounds better.
* It switched between "Dashboards" and "dashboards" . The uppercase makes more sense to me in this context because when referencing other plugins I see you are lower casing them.
* Developers can also configure this to start without modifying the yaml file by starting it up with `data_source.enabled=true`. Since MDS is also a core osd feature we could potentially add a new script in the package.json file like `yarn start:mds` which essentially just calls start but with data source enabled.

Coupling that with my comment about updating the cli for generating a plugin then could be really easy for developers to get this.

Awesome though. 😎

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,258 @@
+categories:
+ - technical-post
+meta_keyword: OpenSearch multiple data source, OpenSearch Dashboards
+meta_description: This blog provides a guideline on howw integrate multiple data source with Dashboards pluginds
```
Should be plugins

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,258 @@
+
+**Initialize and utilize multiple data source clients to make API calls**
+
+Currently, OpenSearch Dashboards allows two types of clients in use, opensearch-js and legacy client. The OpenSearch client is designed for interacting with OpenSearch clusters while legacy client provides compatibility to work with older version of ElasticSearch. 
```
Sure i will modify this section to make it more general and understandable

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,258 @@
+
+*1.5.1 Client side to register dataSourceManagement*
+
+For plugin wants to integrate with multiple data source, data source management offers various component interfaces for choosing data sources. These interfaces include single selection, multiple selection, and readonly view options.  To utilize data source UI component interfaces, the dashboards plugin must initialize dataSourceManagement during the initialization process. Subsequently, in the la ...
```
Got this

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,258 @@
+## 4 FAQ 
+
+Is any UI component provided by dataSourceManagemet used in core?
+Data source selector is now used in devtools, tutorial, saved object management page. You can see the component after enabling data_source.enabed in opensearch_dashbards.yml.
```
Got this

zhyuanqi commented:
> Looks great ! Couple comments but really just nitpicks. But I don't have the best grammar skills.
> 
> General nitpicks:
> 
> * I noticed it switches between "multiple data source" and "multiple data sources" it think the latter sounds better.
> * It switched between "Dashboards" and "dashboards" . The uppercase makes more sense to me in this context because when referencing other plugins I see you are lower casing them.
> * Developers can also configure this to start without modifying the yaml file by starting it up with `data_source.enabled=true`. Since MDS is also a core osd feature we could potentially add a new script in the package.json file like `yarn start:mds` which essentially just calls start but with data source enabled.
> 
> Coupling that with my comment about updating the cli for generating a plugin then could be really easy for developers to get this.
> 
> Awesome though. 😎

Thanks. Let me change to "multiple data sources" and "Dashboards"

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+*1.5.2 How to use UI component interfaces provide by dataSourceManagement plugin*
+
+Now that we've initialized the `dataSourceManagement` plugin, it is time to use the component interfaces it provides.  Using selector as an example below, it used OUI combobox as base component and provides a user interface for selecting data sources. It is stateful and contains the logic to fetch available data sources, return user selection back to user via callback function, and update the UI  ...
```
```suggestion
Now that you've initialized the `dataSourceManagement` plugin, you can use the component interfaces it provides.  The example selector below uses OUI combobox as a base component and provides a UI for selecting data sources. It is stateful and contains the logic to fetch available data sources, return the user selection back to the user through a callback function, and update the UI elements accordingly.
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,341 @@
+const [selectedDataSources, setSelectedDataSources] = useState<string[]>([]);
+```
+
+### 1.6 Step 6: run and see a working example
````
```suggestion
### 1.6 Step 6: Run a working example
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+### 1.6 Step 6: run and see a working example
+
+Now that we have everything ready. Let’s combine it to a working example plugins! Let’s first added a datasource selector in app.tsx.
```
```suggestion
Now that everything is ready, you can combine it into a working example plugin! First add a data source selector in app.tsx:
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+Get indices from local cluster:
+<img src="/assets/media/blog-images/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins/blog_example_local_cluster.png">
+
+For additional details on the example plugin code, please refer to this [repo](https://github.com/zhyuanqi/OpenSearch-Dashboards/tree/blog_example/examples/blog_example_plugin). 
```
```suggestion
For additional details on the example plugin code, please refer to [this repo](https://github.com/zhyuanqi/OpenSearch-Dashboards/tree/blog_example/examples/blog_example_plugin). 
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+## 3 FAQ 
+
+Is any UI component provided by dataSourceManagemet used in core?
+Data source selector is now used in devtools, tutorial, saved object management page. You can see the component after setting `data_source.enabled` to true in `opensearch_dashbards.yml`.
```
```suggestion
A: A data source selector is now used in Dev Tools, the tutorial, and on the saved object management page. You can see the component after setting `data_source.enabled` to true in `opensearch_dashbards.yml`.
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+Is any UI component provided by dataSourceManagemet used in core?
+Data source selector is now used in devtools, tutorial, saved object management page. You can see the component after setting `data_source.enabled` to true in `opensearch_dashbards.yml`.
+
+Given that a dashboard is a saved object itself, I assume it will not have a data connection associated with it?
```
```suggestion
Q: Given that a dashboard is a saved object itself, will it not have a data connection associated with it?
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+Data source selector is now used in devtools, tutorial, saved object management page. You can see the component after setting `data_source.enabled` to true in `opensearch_dashbards.yml`.
+
+Given that a dashboard is a saved object itself, I assume it will not have a data connection associated with it?
+A data source, represented as a saved object, is identified by its unique saved object ID. For saved objects created from data sources, it has a references section with ID, name and type. This reference helps us to identify the associated data source for the saved object. Here is an example of a index-pattern reference.
```
```suggestion
A: A data source, represented as a saved object, is identified by its unique saved object ID. Saved objects created from data sources have a references section containing the ID, name, and type. This reference helps you to identify the associated data source for the saved object. Here is an example of a index pattern reference:
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,341 @@
+    }
+]
+```
+The visualization reference is slightly different as it points to the related index-pattern and use the index-pattern id to find corresponding data source. 
````
```suggestion
The visualization reference is slightly different because it points to the related index pattern and uses the index pattern ID to find the corresponding data source: 
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,341 @@
+```
+
+
+## 4 List of supported Plugins/Feature
````
```suggestion
## 4 List of supported plugins/features
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+## 4 List of supported Plugins/Feature
+
+Besides native multiple datasource support for Discover App, Dashboards, Visualization, Dev Tool, Example Data. Starting 2.14, we have enabled multiple datasource in 9 external Dashboards plugins: 
```
```suggestion
Starting in version 2.14, multiple data sources are supported in nine external Dashboards plugins: 
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+Besides native multiple datasource support for Discover App, Dashboards, Visualization, Dev Tool, Example Data. Starting 2.14, we have enabled multiple datasource in 9 external Dashboards plugins: 
+
+* index management
```
```suggestion
* Index Management
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+Besides native multiple datasource support for Discover App, Dashboards, Visualization, Dev Tool, Example Data. Starting 2.14, we have enabled multiple datasource in 9 external Dashboards plugins: 
+
+* index management
+* machine learning
```
```suggestion
* Machine Learning
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+* index management
+* machine learning
+* search relevance
```
```suggestion
* Search Relevance
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* index management
+* machine learning
+* search relevance
+* anomaly detection
```
```suggestion
* Anomaly Detection
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* machine learning
+* search relevance
+* anomaly detection
+* maps
```
```suggestion
* Maps
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* search relevance
+* anomaly detection
+* maps
+* security
```
```suggestion
* Security
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* anomaly detection
+* maps
+* security
+* notification
```
```suggestion
* Notification
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* maps
+* security
+* notification
+* query workbench
```
```suggestion
* Query Workbench
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* security
+* notification
+* query workbench
+* trace-analytics 
```
```suggestion
* Trace Analytics 
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* query workbench
+* trace-analytics 
+
+Additionally, we've extended this functionality to 2 core visualizations: Time Series Visual Builder (TSVB) and Region Map as part of Map Plugin. We are also thrilled about further expanding support for multiple data sources across additional plugins and incorporate more features in the future!
```
```suggestion
Additionally, we've extended this functionality to two core visualizations: Time Series Visual Builder (TSVB) and Region Map.
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+## 5 Conclusion
+
+In this blog, we've covered the process of integrating OpenSearch Dashboards plugins with multiple data sources. We strongly encourage you to explore and experiment with multiple data sources using the example plugins provided above. Thanks to Lu Yu  for developing the example plugin! This exploration will enable you to experience the benefits of accessing multiple data sources within a single da ...
```
```suggestion
In this blog post, we've covered the process of integrating OpenSearch Dashboards plugins with multiple data sources. We strongly encourage you to experiment with multiple data sources using the example plugins provided above. Thanks to Lu Yu for developing the example plugin! This exploration will enable you to experience the benefits of accessing multiple data sources within a single dashboard as well as allow your plugin to expose a wider range of data from remote clusters.
```

natebower reviewed (commented):
@zhyuanqi @pajuric Please see my comments and changes and let me know if you have any questions. I had a bit of difficulty guessing at the proper articles, and I'm not totally confident that my changes retained the intended technical meaning in all cases, so I would recommend having a technical writer review for accuracy. Thanks!

natebower commented:
Per @brijos, he will work with the appropriate technical writer to revise the draft (as necessary) and then send for a second editorial review.

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+### 1.2 Step 2: Configure plugin to add dependencies
+
+In the opensarch_dashboards.json file of your Dashboards plugin, add data source and data source management in optional plugins section. The data source plugin provides the clients to connect with multiple data sources. Data source management plugin provides generic interfaces to consume data source components such as data source menu and/or data source selector to be used on pages depending on s ...
```
sure

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+**Server side: Verify if multiple data sources is enabled and register plugin API schema accordingly**
+
+If your Dashboards plugin depends on a new API provided by your new server side plugin other than OpenSearch APIs, you can register the schema for the API or pass the request path as parameters when initializing the request. 
```
Got it

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+### 1.5 Step 5: Register datasource management plugin under public folder
+
+*1.5.1 Extract dataSourceManagement object in plugin module under public folder*
```
sure

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,341 @@
+yarn start --run-examples
+```
+
+Once Dashboards are up and running, navigate to the page and click on the left-side menu. From there, open the Developer examples under OpenSearch Dashboards. Then click on Multiple Data Sources Integration. On the right side of the page, you'll find a comprehensive list of pickers, each accompanied by a brief introduction and a set of properties listed below. Feel free to explore and experiment  ...
````
Got it

seraphjiang commented:
> Per @brijos, he will work with the appropriate technical writer to revise the draft (as necessary) and then send for a second editorial review.

Hi @natebower is there any update. could we are able to close this this week?

hdhalter commented:
I can have a tech writer review it. In the meantime, @zhyuanqi, can you address the style errors? Thanks!

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+#### 1.5.1 Extract the `dataSourceManagement` object in the plugin module in a public folder
+
+The Data Source Management plugin offers various react components through a generic interface for use by your Dashboards plugin. To use data source UI component interfaces, extract the `dataSourceManagement` object in the plugin module in a public folder and pass it to the main application: 
```
```suggestion
The Data Source Management plugin provides react components through a generic interface for your OpenSearch Dashboards plugin. To use the data source UI component interfaces, extract the `dataSourceManagement` object from the plugin module in a public folder and pass it to the main application: 
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,341 @@
+}
+```
+
+#### 1.5.2 How to use UI component interfaces provide by the `dataSourceManagement` plugin
````
```suggestion
#### 1.5.2 How to use UI component interfaces provide by the `dataSourceManagement` plugin
```
```suggestion
Now, use the UI component interfaces provided by the `dataSourceManagement` plugin.
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+<img src="/assets/media/blog-images/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins/blog_example_remote_cluster.png">
+
+Get indexes from a local cluster:
+<img src="/assets/media/blog-images/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins/blog_example_local_cluster.png">
```
Deleted for redundancy

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+The data source is persisted as a saved object in the Dashboards system index. At the time of creation, an UUID is assigned as its unique identifier, meta data of the data source is also gathered including remote cluster version, and associated plugins. These autogenerated data as well as user provided information including its title, endpoint, authentication methods, are stored as attributes of  ...
+
+## 3 FAQ 
```
+1

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+<img src="/assets/media/blog-images/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins/multi_selectable.png">
+
+### 2.2 The data source as a saved object
```
```suggestion
### Data sources as saved objects
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+### 2.2 The data source as a saved object
+
+The data source is persisted as a saved object in the Dashboards system index. At the time of creation, a UUID is assigned as its unique identifier and data source metadata is also gathered, including the remote cluster version and associated plugins. This automatically generated data, as well as user-provided information including the title, endpoint, and authentication methods, are stored as at ...
```
```suggestion
The data source is persisted as a saved object in the OpenSearch Dashboards system index. Upon creation, it is assigned a UUID as a unique identifier, and metadata such as the remote cluster version and associated plugins is gathered. This automatically generated data, along with user-provided information like the title, endpoint, and authentication methods, is stored as attributes of the data source object. Refer to the [related code](https://github.com/opensearch-project/OpenSearch-Dashboards/blob/main/src/plugins/data_source/common/data_sources/types.ts#L8) for the current object schema. 
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+The data source is persisted as a saved object in the Dashboards system index. At the time of creation, a UUID is assigned as its unique identifier and data source metadata is also gathered, including the remote cluster version and associated plugins. This automatically generated data, as well as user-provided information including the title, endpoint, and authentication methods, are stored as at ...
+
+## 3 FAQ 
```
```suggestion
## Frequently asked questions 
```

vagimeli replied:
I suggest moving the FAQs to after the plugin list section.

zhyuanqi replied:
Got it.

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+## 3 FAQ 
+
+Q: Is any UI component provided by `dataSourceManagemet` used in core OpenSearch?
+A: A data source selector is now used in Dev Tools, the tutorial, and on the saved object management page. You can see the component after setting `data_source.enabled` to true in `opensearch_dashbards.yml`.
```
```suggestion
A: A data source selector is now used in Dev Tools, the tutorial, and on the saved object management page. You can see the component after setting `data_source.enabled` to `true` in your `opensearch_dashbards.yml` file.
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+Q: Is any UI component provided by `dataSourceManagemet` used in core OpenSearch?
+A: A data source selector is now used in Dev Tools, the tutorial, and on the saved object management page. You can see the component after setting `data_source.enabled` to true in `opensearch_dashbards.yml`.
+
+Q: Given that a dashboard is a saved object itself, will it not have a data connection associated with it?
```
```suggestion
Q: Given that a dashboard is a saved object itself, will it not have a data connection associated with it?
```
```suggestion
Q: Since a dashboard is a saved object, won't it have an associated data connection?
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+A: A data source selector is now used in Dev Tools, the tutorial, and on the saved object management page. You can see the component after setting `data_source.enabled` to true in `opensearch_dashbards.yml`.
+
+Q: Given that a dashboard is a saved object itself, will it not have a data connection associated with it?
+A: A data source, represented as a saved object, is identified by its unique saved object ID. Saved objects created from data sources have a references section containing the ID, name, and type. This reference helps you to identify the associated data source for the saved object. Here is an example of a index pattern reference:
```
```suggestion
A: A data source, represented as a saved object, is identified by its unique saved object ID. Saved objects created from data sources have a `references` section containing the ID, name, and type. This reference helps identify the associated data source for the saved object. Here's an example of an index pattern reference:
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,341 @@
+```
+
+
+## 4 List of supported plugins/features
````
```suggestion
## List of supported plugins/features
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+## 4 List of supported plugins/features
+
+Starting in version 2.14, multiple data sources are supported in nine external Dashboards plugins: 
```
```suggestion
Starting from version 2.14, the following external OpenSearch Dashboards plugins support multiple data sources: 
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+* Query Workbench
+* Trace Analytics 
+
+Additionally, we've extended this functionality to two core visualizations: Time Series Visual Builder (TSVB) and Region Map.
```
```suggestion
Multiple data sources functionality is also now available with Time Series Visual Builder (TSVB) and Region Map visualization types.
```

vagimeli replied:
Should we also include Vega? See PR https://github.com/opensearch-project/project-website/pull/2837. FYI--user documentation is in process for this topic.

zhyuanqi replied:
Got it

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+Additionally, we've extended this functionality to two core visualizations: Time Series Visual Builder (TSVB) and Region Map.
+
+
+## 5 Conclusion
```
```suggestion
## Conclusion
```

vagimeli commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+
+## 5 Conclusion
+
+In this blog post, we've covered the process of integrating OpenSearch Dashboards plugins with multiple data sources. We strongly encourage you to experiment with multiple data sources using the example plugins provided above. Thanks to Lu Yu for developing the example plugin! This exploration will enable you to experience the benefits of accessing multiple data sources within a single dashboard  ...
```
```suggestion
This blog post covered the process of integrating OpenSearch Dashboards plugins with multiple data sources. We encourage you to experiment with multiple data sources using the provided example plugins, developed by Lu Yu. This will allow you to experience the benefits of accessing multiple data sources within a single dashboard and enable your plugin to expose a wider range of data from remote clusters.
```

vagimeli reviewed (commented):
Doc review complete. Please see edits and comments. Once doc review comments are addressed, this can move forward to editorial review. Note that the editor is OOO until Tuesday, May 21.

pajuric commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+date: 2024-05-01
+categories:
+ - technical-post
+meta_keyword: OpenSearch multiple data sources, OpenSearch Dashboards
```
@zhyuanqi - please update the blog with the following meta: 

meta_keywords:  multiple data sources in OpenSearch, OpenSearch Dashboards plugins, using multiple data sources 

meta_description: Leverage this comprehensive guide to learn how you can integrate multiple multiple data sources in OpenSearch to create visually compelling dashboards using plugins.

zhyuanqi replied:
got it

pajuric commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,341 @@
+title: "Developer guideline: Multiple data sources in OpenSearch Dashboards and plugins"
+authors:
+ - zhyuanqi
+date: 2024-05-01
```
Please update the date to 5/21/2024

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,336 @@
+yarn start --run-examples
+```
+
+With OpenSearch Dashboards running, navigate to the page and select the left menu. Under OpenSearch Dashboards, open **Developer Examples**, then select **Multiple Data Sources Integration**, as shown in the following image. On the right side, you'll find a list of pickers, each with a brief introduction and properties listed below. Feel free to experiment with the example plugin.
````
```suggestion
With OpenSearch Dashboards running, navigate to the page and select the menu on the left. Under OpenSearch Dashboards, open **Developer Examples** and then select **Multiple Data Sources Integration**, as shown in the following image. On the right, you'll find a list of pickers, each with a brief introduction and a list of properties. Feel free to experiment with the example plugin.
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,336 @@
+yarn start --run-examples
+```
+
+With OpenSearch Dashboards running, navigate to the page and select the left menu. Under OpenSearch Dashboards, open **Developer Examples**, then select **Multiple Data Sources Integration**, as shown in the following image. On the right side, you'll find a list of pickers, each with a brief introduction and properties listed below. Feel free to experiment with the example plugin.
````
Second sentence: Either "Under **OpenSearch Dashboards**, or the sentence should just start with "Open **Developer Examples**..."

zhyuanqi replied:
change to `Open Developer examples, ...`

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+
+## List of supported plugins/features
+
+Starting from version 2.14, the following external OpenSearch Dashboards plugins support multiple data sources: 
```
```suggestion
Starting in version 2.14, the following external OpenSearch Dashboards plugins support multiple data sources: 
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+* Anomaly Detection
+* Maps
+* Security
+* Notification
```
```suggestion
* Notifications
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+* Query Workbench
+* Trace Analytics 
+
+Multiple data sources functionality is also now available with Time Series Visual Builder (TSVB), Vega and Region Map visualization types.
```
```suggestion
Multiple data sources functionality is also now available with the Time Series Visual Builder (TSVB), Vega, and Region Map visualization types.
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+## Frequently asked questions 
+
+Q: Is any UI component provided by `dataSourceManagemet` used in core OpenSearch?
+A: A data source selector is now used in Dev Tools, the tutorial, and on the saved object management page. You can see the component after setting `data_source.enabled` to `true` in your `opensearch_dashbards.yml` file.
```
Should "saved object management" be formatted as a UI element (the name of the page)?

zhyuanqi replied:
got it

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+Q: Is any UI component provided by `dataSourceManagemet` used in core OpenSearch?
+A: A data source selector is now used in Dev Tools, the tutorial, and on the saved object management page. You can see the component after setting `data_source.enabled` to `true` in your `opensearch_dashbards.yml` file.
+
+Q: Since a dashboard is a saved object, won't it have an associated data connection?
```
```suggestion
Q: Because a dashboard is a saved object, won't it have an associated data connection?
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,336 @@
+    }
+]
+```
+The visualization reference is slightly different as it points to the related index pattern and uses the index pattern ID to find the corresponding data source, as shown in the following example:
````
```suggestion
The visualization reference is slightly different because it points to the related index pattern and uses the index pattern ID to find the corresponding data source, as shown in the following example:
```

natebower commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+
+## Conclusion
+
+This blog post covered the process of integrating OpenSearch Dashboards plugins with multiple data sources. We encourage you to experiment with multiple data sources using the provided example plugins, developed by Lu Yu. This will allow you to experience the benefits of accessing multiple data sources within a single dashboard and enable your plugin to expose a wider range of data from remote cl ...
```
Should we include Lu Yu's GitHub handle?

zhyuanqi replied:
I have provided how to install example plugins above

natebower reviewed (commented):
@zhyuanqi @vagimeli Please see my comments and changes and let me know if you have any questions. Thanks so much for all of your work on this! @pajuric This will be ready for the publishing process once the author indicates that my comments/changes have been addressed.

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+
+Now, use the UI component interfaces provided by the `dataSourceManagement` plugin.
+
+With the `dataSourceManagement` plugin initialized, you can leverage its component interfaces. The following example demonstrates using the OpenSearch UI combobox as a base component to provide a UI for selecting data sources. It is stateful, fetching available data sources, returning the user's selection through a callback function, and updating the UI accordingly.
```
Should be the user's/ User do the selection, not developer

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,336 @@
+)
+```
+
+Congratulations! Now you can run your plugins and use the pickers to select the desired data source. 
````
SHould be select desired data source as we do not provide mult-selection

natebower commented:
@zhyuanqi It looks like there are still comments/suggestions in this PR that will need to be resolved before this can be moved forward in the process. Just wanted to call your attention to it so that this doesn't fall through the cracks 😄.

zhyuanqi commented:
> @zhyuanqi It looks like there are still comments/suggestions in this PR that will need to be resolved before this can be moved forward in the process. Just wanted to call your attention to it so that this doesn't fall through the cracks 😄.

Okay, Let me check on it.

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+
+### Step 3: Add the data source parameter to server-side plugins
+
+**Server side:** Verify multiple data source enablement and register the plugin API accordingly.
```
Got it

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
````diff
@@ -0,0 +1,336 @@
+    }
+  };
+```
+Now, combine the preceding function and react component:
````
Got it

zhyuanqi commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+
+Extract the `dataSourceManagement` object in the plugin module in a public folder.
+
+The Data Source Management plugin provides react components through a generic interface for your OpenSearch Dashboards plugin. To use the data source UI component interfaces, extract the `dataSourceManagement` object from the plugin module in a public folder and pass it to the main application: 
```
Got it

zhyuanqi commented:
@natebower  I should have applied all the comments in the newest commit. Could you help to check if anything missing

natebower commented:
> @natebower I should have applied all the comments in the newest commit. Could you help to check if anything missing

@zhyuanqi It looks like there is just one comment outstanding on line 310, but otherwise, LGTM.

@pajuric Confirmed that all other comments/suggestions appear to have been addressed and that this should be good to go once the remaining one is resolved. Thanks!

zhyuanqi commented:
> > @natebower I should have applied all the comments in the newest commit. Could you help to check if anything missing
> 
> @zhyuanqi It looks like there is just one comment outstanding on line 310, but otherwise, LGTM.
> 
> @pajuric Confirmed that all other comments/suggestions appear to have been addressed and that this should be good to go once the remaining one is resolved. Thanks!

I have addressed the comment on line 310. Thanks for checking on my doc

pajuric commented:
@nateynateynate @krisfreedain - Blog is ready to publish tomorrow - Tuesday, June 4. Thanks.

krisfreedain commented on _posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md:
```diff
@@ -0,0 +1,336 @@
+layout: post
+title: "Developer guide: Integrating multiple data sources using OpenSearch Dashboards and plugins"
+authors:
+ - zhyuanqi
```
there doesn't seem to be a profile created for 'zhyuanqi' - can you resolve this? 

(would be an appropriate `zhyuanqi.md` file under `_community_members`)

krisfreedain replied:
Also, got the following error when I ran this locally:

Liquid syntax error (line 245): Expected end_of_string but found colon in "{{ data: indices ? indices : 'Unknown' }}" in /_posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md

zhyuanqi replied:
Hi Kris. I have added the author part/ For `Liquid syntax error (line 245): Expected end_of_string but found colon in "{{ data: indices ? indices : 'Unknown' }}" in /_posts/2024-05-01-develop-guideline-multiple-data-source-in-opensearch-and-plugins.md`, the `{ data: indices ? indices : 'Unknown' }}` is in part of the code block.

zhyuanqi replied:
added

krisfreedain reviewed (dismissed):
Missing author's Community Member profile.  See notes in review

nateynateynate reviewed (approved):
Looks good! Thanks!
